File: C/randomized_quick_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int randomPartition(int arr[], int low, int high) {
    int random = low + rand() % (high - low + 1);
    swap(&arr[random], &arr[high]);

    int pivot = arr[high], i = low - 1;
    for (int j = low; j < high; j++)
        if (arr[j] < pivot)
            swap(&arr[++i], &arr[j]);

    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

void randomizedQuickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = randomPartition(arr, low, high);
        randomizedQuickSort(arr, low, pi - 1);
        randomizedQuickSort(arr, pi + 1, high);
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    srand(time(0)); 

    int arr[] = {10, 80, 30, 90, 40, 50, 70};
    int n = sizeof(arr) / sizeof(arr[0]);

    randomizedQuickSort(arr, 0, n - 1);
    printf("Sorted array:\n");
    printArray(arr, n);

    return 0;
}

File: Lua/randomized_quick_sort.lua
function partition(arr, left, right)
    local pivot = arr[right]
    local i = left - 1
    for j = left, right - 1 do
        if arr[j] <= pivot then
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]
        end
    end
    arr[i + 1], arr[right] = arr[right], arr[i + 1]
    return i + 1
end

function randomized_partition(arr, left, right)
    local rand_index = math.random(left, right)
    arr[right], arr[rand_index] = arr[rand_index], arr[right]
    return partition(arr, left, right)
end

function randomized_quick_sort(arr, left, right)
    if left < right then
        local pi = randomized_partition(arr, left, right)
        randomized_quick_sort(arr, left, pi - 1)
        randomized_quick_sort(arr, pi + 1, right)
    end
end

math.randomseed(os.time())
local arr = {5, 1, 4, 2, 8}
randomized_quick_sort(arr, 1, #arr)
print(table.concat(arr, " "))
